protocol/standard-l2-genesis.md

```diff
@@ -173,14 +230,24 @@ to source this config, it would likely need to be hardcoded in the binary and th
 <!-- An overview of what could go wrong.
 Also any open questions that need more work to resolve. -->
 
+## Sequenced transactions on a fresh chain
+
 There is a concern that the sequencer can include transactions before these values are set on L2.
 If we define the `SystemConfig.startBlock` as the [first block to start derivation in](https://github.com/ethereum-optimism/optimism/blob/d05fb505809717282d5cee7264a09d26002a4ddd/op-node/cmd/genesis/cmd.go#L174C30-L174C40),
 which is set on `SystemConfig.initialize` and also the deposit transactions that set these values are
 sent in the same block, then we should have the guarantee that no user transactions are included before
 the deposit transactions.
 
+## Max Resource Limit on Deposit Transactions
+
 There is a concern around the max deposit gas limit being too small so that the `SystemConfig` cannot
 deposit all of these values, we should have logic that reverts if the deposit gas limit is too small
-in the `ResourceConfig` sanity check function. Since the `ResourceConfig` can be modified during
+in the `setResourceConfig()` function's [sanity checks](https://github.com/ethereum-optimism/optimism/blob/feat/holocene-contracts/packages/contracts-bedrock/src/L1/SystemConfig.sol#L538-L556). Since the `ResourceConfig` can be modified during
 `initialize`, its not that big of a deal and chain operators will see that they need to increase that
 value.
+
+A related concern is that the `useGas()`
+[function](https://github.com/ethereum-optimism/optimism/blob/f99424ded3917ddc0c4ef14355d61e50a38d4d0d/packages/contracts-bedrock/src/L1/ResourceMetering.sol#L156)
+which is called in the `upgrade()` and `setConfig()` functions does not check that the max resource limit is not exceeded by
+the additional `prevBoughtGas`. This is in contrast with
+[`_metered()`](https://github.com/ethereum-optimism/optimism/blob/f99424ded3917ddc0c4ef14355d61e50a38d4d0d/packages/contracts-bedrock/src/L1/ResourceMetering.sol#L128C1-L132C10) which does check `prevBoughtGas`. This requires investigation.
```

Comment:
I think this opens up the following attack: 
1. The system config is reinitialized, which causes it to submit a large amount of data to L2 (we don't yet know how much L2 gas will be required, but it could be that in the future it grows to be significant). 
2. An attacker observes this transaction, and frontruns it to fill up the resource limit with a large deposit tx.
3. The attacker's transaction results in a deposit block larger than the L2 block gas limit.

Reply:
I think the fix is to just add a require statement into `useGas()`. If config deposits are being DoS'd, we can just submit via flashbots

Reply:
There is a buffer between the L1 deposit gas limit and the L2 gas limit. As long as the initializer deposits don't consume more than that buffer, then we are safe. Right now I believe the buffer is 9million and will increase when we increase the L2 block gas limit. With the deposit queue, this problem goes away since it will result in deposits being "dripped" into L2

Reply:
We could put a require statement inside of `useGas` that takes into account the buffer, this method would not be dos'able via front run since calling `useGas` can only be done by a chain operator/governance

Reply:
Agree the deposit-queue is the right long term fix, but for now we just need to work with what we've got to find a good enough solution.

> We could put a require statement inside of `useGas` that takes into account the buffer, this method would not be dos'able via front run since calling `useGas` can only be done by a chain operator/governance

Since `useGas()` and `_metered` both modify `prevBoughtGas`, I had envisioned reusing the [existing check](https://github.com/ethereum-optimism/optimism/blob/f99424ded3917ddc0c4ef14355d61e50a38d4d0d/packages/contracts-bedrock/src/L1/ResourceMetering.sol#L156):

```
if (int256(uint256(params.prevBoughtGas)) > int256(uint256(config.maxResourceLimit))) {
    revert OutOfGas();
}
```

This would indeed be dos'able by front-running the chain-operator with a large deposit tx. 

> There is a buffer between the L1 deposit gas limit and the L2 gas limit. As long as the initializer deposits don't consume more than that buffer, then we are safe. Right now I believe the buffer is 9million and will increase when we increase the L2 block gas limit. With the deposit queue, this problem goes away since it will result in deposits being "dripped" into L2. 

There are two calls to `useGas()` to consider:
2. in `OptimismPortal.setConfig()`: this uses 200_000 (`SYSTEM_DEPOSIT_GAS_LIMIT`) gas per call
1. in `OptimismPortal.upgrade()`: these calls will also be fairly cheap as they just need to overwrite the implementation slot.

Having outlined that, there appears to be very little danger of exceeding the buffer, so I am in favor of not adding the check, but I will look for an off-chain testing approach to ensure we don't exceed the buffer in the future.

Reply:
Tracking here: https://github.com/ethereum-optimism/optimism/issues/12578#issue-2606361806

Reply:
Tracked in this issue: https://github.com/ethereum-optimism/optimism/issues/12578